[FLINK-33565][Scheduler] ConcurrentExceptions works with exception merging

1996fanrui opened the pull request:
## What is the purpose of the change

The concurrent exceptions doesn't work.


## Brief change log

- [FLINK-33565][Exception] Archive exceptions into the exception history immediately when they occur, instead of archiving them when restarting
  - The first commit is refactoring. 
  - Currently, archiving exception when restarting task instead of immediately. It means, when one task failure, we can see the exception history after flink restart this task. 
  - So the first commit is only a refactoring. It archives exceptions into the exception history immediately when they occur, instead of archiving them when restarting.
- [FLINK-33565][Exception] Restart strategy checks whether current failure is a new attempt
  - The second commit is related to restart strategy, adding a return value indicates whether current failure is a new attempt.
- [FLINK-33565][Scheduler] ConcurrentExceptions works with exception merging
  - The third commit is core solution of this JIRA:
  - If it's a new attempt, it's root exception. and it will be the latest root exception.
  - If it's not a new attempt, it will be a concurrentException and it will be added to the latest RootException.
- [FLINK-33565][Scheduler] Correct the numberOfRestarts metric
  - numberOfRestarts should be increased when exception is a new attempt
- [FLINK-33565][Scheduler][refactor] Remove the wrong `@Nullable` annotation of FailureHandlingResult due to the constructor checks `verticesToRestart` is not null




## Verifying this change

- Improve some tests of `ExecutionFailureHandlerTest`, and added new test: `testNewAttemptAndNumberOfRestarts`
- Improve some old tests of `ExponentialDelayRestartBackoffTimeStrategyTest` to check whether `isNewAttempt` is expected
- Improve some tests of `FailureHandlingResultTest` to test `isNewAttempt` are true and false
- Improve the `DefaultSchedulerTest#testExceptionHistoryConcurrentRestart`
- Check the `isNewAttempt` is expected in `FailureHandlingResultSnapshotTest`
- Test `addConcurrentExceptions` in `RootExceptionHistoryEntryTest`

### Test manually

I added a job demo with 6 regions, all tasks will fail when processing the first record, this demo job can be run directly. Here is the result, we can see all failed tasks in the WebUI.

![image](https://github.com/apache/flink/assets/38427477/8830b0eb-aaef-4ef6-b6dd-888f496aa229)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper:no
  - The S3 file system connector:no

## Documentation

  - Does this pull request introduce a new feature? no

1996fanrui reviewed (commented):
Hi @XComp , would you mind helping take a look in your free time? Thanks~

 I didn't finish the detailed unit test due to I wanna check with you whether the solution is fine. The test manually is fine.

![image](https://github.com/apache/flink/assets/38427477/fab2369d-48b6-42b2-b32f-45e73533a020)

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -703,27 +707,41 @@ private void archiveGlobalFailure(
+            checkState(latestRootExceptionEntry != null, "It should have old failure.");
+            List<Execution> concurrentlyExecutions = new LinkedList<>();
+            failureHandlingResult.getRootCauseExecution().ifPresent(concurrentlyExecutions::add);
+            concurrentlyExecutions.addAll(failureHandlingResult.getConcurrentlyFailedExecution());
```
nit: Here we use a LinkedList on a Set (which is returned by `getConcurrentlyFailedExecution`). Either we switch to `ArrayList` to benefit from the `addAll` call or we revert the return type of `getConcurrentlyFailedExecution` back to `Iterable` and benefit from the single-element add performance of `LinkedList`. WDYT? :thinking:

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:
```diff
@@ -150,7 +154,11 @@ public long getTimestamp() {
      * @return The concurrently failed {@code Executions}.
      */
-    public Iterable<Execution> getConcurrentlyFailedExecution() {
+    public Set<Execution> getConcurrentlyFailedExecution() {
```
I'm not sure whether this change is necessary. See more other related comment.

XComp reviewed (commented):
Thanks for your efforts @1996fanrui. Overall, looks good. I can do a proper review after you've finalized the PR. I just added a few nitty comments. PTAL

1996fanrui commented:
> Thanks for your efforts @1996fanrui. Overall, looks good. I can do a proper review after you've finalized the PR. I just added a few nitty comments. PTAL

Many thanks @XComp for the review! If the solution is fine, I will address your comments and finish the tests as soon as possible!

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java:
```diff
@@ -367,17 +367,13 @@ private void restartTasksWithDelay(final FailureHandlingResult failureHandlingRe
                         FutureUtils.assertNoException(
                                 cancelFuture.thenRunAsync(
-                                        () -> {
-                                            archiveFromFailureHandlingResult(
```
Thanks for the clarification!

The initial motivation might for collecting all concurrent exceptions. In this PR, the solution is save the `latestRootExceptionEntry` as a filed in `SchedulerBase`, when all subsequent non-root exceptions will be added to the `latestRootExceptionEntry`.

1996fanrui reviewed (commented):
Thanks @XComp  for the suggestion again!

I have addressed all comments, and added some tests.

1996fanrui commented:
Hi @XComp , sorry for bother you again due to the 1.19 will be freeze next week.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExecutionFailureHandlerTest.java:
```diff
@@ -71,7 +74,10 @@ void setUp() {
+        isNewAttempt = new AtomicBoolean(true);
+        backoffTimeStrategy =
+                new TestRestartBackoffTimeStrategy(
+                        true, RESTART_DELAY_MS, () -> isNewAttempt.get());
```
```suggestion
                        true, RESTART_DELAY_MS, isNewAttempt::get);
```
nit

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExecutionFailureHandlerTest.java:
```diff
@@ -171,6 +179,9 @@ void testNonRecoverableFailureHandlingResult() throws Exception {
         assertThat(result.getFailureLabels().get())
                 .isEqualTo(testingFailureEnricher.getFailureLabels());
         assertThat(result.getTimestamp()).isEqualTo(timestamp);
+        // NonRecoverableFailure is new attempt even if RestartBackoffTimeStrategy consider it's not
+        // new attempt.
+        assertThat(result.isNewAttempt()).isTrue();
```
```suggestion
        assertThat(result.isNewAttempt())
                .as(
                        "A NonRecoverableFailure should be new attempt even if RestartBackoffTimeStrategy consider it's not new attempt.")
                .isTrue();
```
just another nitty hint: For tests, we don't need to add comments but could rather have a description assertion message. That serves the same purpose as the command but also makes the test execution more descriptive.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExecutionFailureHandlerTest.java:
```diff
@@ -183,6 +194,55 @@ void testNonRecoverableFailureHandlingResult() throws Exception {
         assertThat(executionFailureHandler.getNumberOfRestarts()).isZero();
     }
 
+    /** Test isNewAttempt of {@link FailureHandlingResult} is expected. */
+    @Test
+    void testNewAttemptAndNumberOfRestarts() throws Exception {
```
The comment is obsolete because it could be expressed through the test method name(s) (to reduce redundant information in the code). 

Additionally, what's your opinion on splitting this big test up into smaller test methods. I see the benefit of having one big one. Because you can check the number of restarts at the same time. But having more specific test methods usually helps with improving readability of the test code.. WDYT?

XComp replied:
Like in my previous comment: We can avoid using comments in favor of the assert message :shrug:

1996fanrui replied:
Good suggestion!

Also, I extracted the `assertHandlerRootException` and `assertHandlerConcurrentException` methods, after that, the `testNewAttemptAndNumberOfRestarts` is totally simple.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntryTest.java:
```diff
@@ -81,10 +81,14 @@ void testFromFailureHandlingResultSnapshot() throws ExecutionException, Interrup
+                getExceptionHistoryEntryPredicate(
+                        concurrentException1, concurrentlyFailedExecutionVertex1);
+
+        final Throwable concurrentException2 = new IllegalStateException("Expected other failure2");
+        final ExecutionVertex concurrentlyFailedExecutionVertex2 = extractExecutionVertex(2);
```
Can you move these two lines further down where they are actually used (alternatively, you could move the Predicate initialization up). The goal would be to have the entire initialization in one place. Me personally, I would prefer moving the code down. That gives less distraction to the reader so that he/she can focus on the first concurrent exception handling at the beginning. But up to you...

XComp replied:
Thinking about it: Another solution would be to create a dedicated test method `testAddConecurrentExceptions`. WDYT? :thinking:

1996fanrui replied:
My thoughts is moving the code down instead of create a new test method.

Because `testAddConecurrentExceptions` will use all code of this test. WDYT?

XComp replied:
Fine with me. About reusing the code: You could move the code that's common between tests into a private static method. But other's might argue that this is also not good for code readability.

Anyway, it's your choice in the end. Most of my comments are discussion points rather than comments where I claim that my proposal is the right one. :-)

1996fanrui replied:
Thanks for the explanation!

> testAddConecurrentExceptions will use all code of this test 

If we have 2 tests, and test1 only call `testCommon`, and test2 call `testCommon` and `testPart2`. It means test2 can cover test1.

2 solutions:
- keep test1 and test2
- Only keep test2

In the current scenario, is it enough for us to only keep test2? Looking forward to your opinion, fine with me as well. 😁

XComp replied:
Fine with me to keep it like that.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntryTest.java:
```diff
@@ -150,25 +164,24 @@ void testFromGlobalFailure() throws ExecutionException, InterruptedException {
                                         || exception1Predicate.test(exceptionHistoryEntry));
     }
 
+    private Predicate<ExceptionHistoryEntry> getExceptionHistoryEntryPredicate(
```
```suggestion
    private Predicate<ExceptionHistoryEntry> triggerFailureAndCreateEntryMatcher(
```
Based on the old name, I didn't expect the triggering of the failure. We might want to mention that in the method name because it's an essential part of each test scenario that the failure is actually registered through a state update in the `ExecutionGraph`

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/FailureHandlingResult.java:
```diff
@@ -206,6 +214,11 @@ public boolean isGlobalFailure() {
     }
 
+    /** @return Whether this failure is a new attempt. */
+    public boolean isNewAttempt() {
```
What about calling it `isRootCause` here analogously to the exception history? ...or isFirstErrorInNewErrorHandlingCycle 🤔 Just as other more descriptive proposals. Up to you, I'm not fighting for any version. :innocent:

XComp reviewed (changes requested):
Thanks for the reminder and sorry for leaving you hanging here for so long. I got distracted by other issues. I went over the PR once more. PTAL

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java:
```diff
@@ -170,11 +176,15 @@ public RootExceptionHistoryEntry(
         this.concurrentExceptions = concurrentExceptions;
     }
 
+    public void addConcurrentExceptions(Iterable<Execution> concurrentlyExecutions) {
```
Good catch, added the `@NotThreadSafe`.

1996fanrui reviewed (commented):
Thanks @XComp for the quick feedback and patient review!

I have addressed all comments~

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/FixedDelayRestartBackoffTimeStrategy.java:
```diff
@@ -61,8 +61,9 @@ public long getBackoffTime() {
-    public void notifyFailure(Throwable cause) {
+    public boolean notifyFailure(Throwable cause) {
         currentRestartAttempt++;
+        return true;
```
I understand that it wasn't covered by FLIP-364. But from a conceptual point of view: shouldn't we cover it also for the two other restart strategies. Or am I missing something here? :thinking:  I'm also find with having this being covered by follow-up Jira issues.

1996fanrui replied:
> But from a conceptual point of view: shouldn't we cover it also for the two other restart strategies. Or am I missing something here?

In the beginning, I want to improve all restart strategies, but I meet some feedback to removing other restart strategies during discussion. The core background is this thread: https://lists.apache.org/thread/l7wyc7pndpsvh2h7hj3fw2td9yphrlox

In brief, 3 reasons:

1. The semantics of option

    - the failure-rate strategy's restart upper limit option is named  `restart-strategy.failure-rate.max-failures-per-interval`
    - It's  `max-failures-per-interval` instead of `max-attempts-per-interval`.
    - If we improve it directly, the name and behaviour aren't matched.
2. We recommend users use the `exponential-delay restart strategy` in the future, it's more powerful.
3. After FLIP-360 discussion, I found `exponential-delay restart strategy` can replace other restart strategies directly if users set the `restart-strategy.exponential-delay.backoff-multiplier` = 1

Actually, I think other restart-strategies can be deprecated in the future.

XComp replied:
Thanks for the clarification. That makes sense.

> After FLIP-360 discussion, I found exponential-delay restart strategy can replace other restart strategies directly if users set the restart-strategy.exponential-delay.backoff-multiplier = 1

I guess, you meant [FLIP-364](https://cwiki.apache.org/confluence/display/FLINK/FLIP-364%3A+Improve+the+exponential-delay+restart-strategy) here? ;-) You're right, we could deprecate the class. But there is still a benefit of keeping the configuration parameter (as an alias that configures the `ExponentialDelayRestartBackoffTimeStrategy` with multiplier=1) to make it easier to configure this behavior. 

What about deprecating that class as part of this PR and coming up with a follow-up Jira issue that replaces the strategy?

1996fanrui replied:
> I guess, you meant [FLIP-364](https://cwiki.apache.org/confluence/display/FLINK/FLIP-364%3A+Improve+the+exponential-delay+restart-strategy) here?

Exactly, sorry for my typo.

> What about deprecating that class as part of this PR and coming up with a follow-up Jira issue that replaces the strategy?

They are internal class, IIUC, we can refactor it directly without the `@Deprecated` annotation.

---------------------------------------------------------------------------------

Your suggestion is that we still keep the failure-rate and fixed-delay restart-strategies, but we can reuse the `ExponentialDelayRestartBackoffTimeStrategy` class, right?

I'm thinking could we deprecate failure-rate and fixed-delay restart-strategies directly? Users can configure them directly.

Keep failure-rate and fixed-delay restart-strategies and reusing `ExponentialDelayRestartBackoffTimeStrategy` class still has the semantic problem that I mentioned in the last comment [1].

[1] https://github.com/apache/flink/pull/24003#discussion_r1456807889

XComp reviewed (commented):
I passed over the changes and left a few minor remarks. We should be close to finalizing it. :)

1996fanrui reviewed (commented):
Thanks @XComp for the quick check! I addressed all your comments again.

1996fanrui commented:
@flinkbot run azure

XComp reviewed (commented):
Looks good to me. Only a few minor naming/typo items left.

XComp reviewed (approved):
LGTM :+1: Thanks for your efforts, @1996fanrui, and for keeping up with my nitty comments. I enjoyed the discussions in this PR. 

About the commits: We can reorganize them a bit, I feel. You could keep https://github.com/apache/flink/pull/24003/commits/6fe8037b40d8458a5524adfb352d96925ced63bd and https://github.com/apache/flink/pull/24003/commits/f422804e918a8ea3537deeb1c4f4f1c9676e1a8a as separate hotfix commits (using the `[hotfix]` prefix rather than the Jira issue) because they improve the code base and we would want to keep them even if we decide to revert FLINK-33565 (for whatever reason) in the future. WDYT?

1996fanrui commented:
> LGTM 👍 Thanks for your efforts, @1996fanrui, and for keeping up with my nitty comments. I enjoyed the discussions in this PR.

Thanks @XComp for the patient review, I definitely learned some skills from your comments. I enjoyed the discussions as well.

> About the commits: We can reorganize them a bit, I feel. You could keep [6fe8037](https://github.com/apache/flink/commit/6fe8037b40d8458a5524adfb352d96925ced63bd) and [f422804](https://github.com/apache/flink/commit/f422804e918a8ea3537deeb1c4f4f1c9676e1a8a) as separate hotfix commits (using the `[hotfix]` prefix rather than the Jira issue) because they improve the code base and we would want to keep them even if we decide to revert [FLINK-33565](https://issues.apache.org/jira/browse/FLINK-33565) (for whatever reason) in the future. WDYT?

The idea makes sense to me. 👍🏻

I update the commit message for the last commit, but  I didn't updated it for the first commit due to some reasons:

1. The first commit changed the behavior
    - Before this PR, archiving exception when restarting task instead of immediately.
    - It means, when one task failure, we can see the exception history after flink restart this task.
    - The first commit archives exceptions into the exception history immediately when they occur, instead of archiving them when restarting.
2. I hope the first commit is fine, but I'm afraid we missed somethings , so it's not a simple refactor or hotfix.

WDYT?

XComp commented:
Fair enough. What about squashing the FLINK-33565-related commits into a single one?

1996fanrui commented:
> Fair enough. What about squashing the [FLINK-33565](https://issues.apache.org/jira/browse/FLINK-33565)-related commits into a single one?

Hi @XComp , it's fine for me. Let me explain why I split them into multiple commits:

1. First of all, the doc of Flink code style introduces us that it's better to separate refactoring, cleanup and independent-changes as the separate commits. As I understand, it has a series of advantages:
  - Easy to review
  - Easy to revert
  - Easy to cherry-pick
    - In my daily work, we always cherry-pick some new features(our users needed) into our internal flink version, because a lot of flink jobs upgrade the flink version needs a lot of efforts.
    - During cherry-pick, I found let the ` independent-changes` as the separate commit is very useful.
    - For example, we need to cherry pick featureB, and featureA is merged before featureB. 
    - FeatureB was using some common changes of FeatureA, FeatureA added some new fields in some common classes.(It's not refactor, but the change isn't huge.)
    - If FeatureA think it's minor change and keep all changed into 1 commit. We must cherry-pick featureA during cherry-pick featureB.
    - If featureA split them into multiple commits for each independent-changes, we can cherry-pick the common class changes of featureA, and the whole featureB.
2. Currently, the first commit changed a minor behavior, and it's similar refactor. And it's mentioned in the last comment: https://github.com/apache/flink/pull/24003#issuecomment-1903758592
3. The second commit, it's `[FLINK-33565][Exception] Restart strategy checks whether current failure is a new attempt`.
    - The concurrent exceptions depend on the result of `RestartBackoffTimeStrategy`.
    - Other features might depend on it as well in the future, so I prefer it's a independent-change
4. The third commit is core change of this feature.
5. The fourth commit is that fixing the `numberOfRestarts` metic, it may be a ndependent-change as well.

That's why I split them into 4 commits. 

As I said before, squashing them into one commit is fine for me. Looking forward to your feedback, thanks~

[1] https://flink.apache.org/how-to-contribute/code-style-and-quality-pull-requests/#3-separate-refactoring-cleanup-and-independent-changes

XComp commented:
ok, fine with me :+1:

1996fanrui commented:
Thanks @XComp for the patient review and a series of great suggestions again! Merging~

A good experience~
